### PR TITLE
Sample for custom widget designer

### PR DESCRIPTION
**Pull request by Gebov**

### CODE REVIEW

File: src/widget-editor/ReadMe.md [hunk 1]
```diff
@@ -0,0 +1,45 @@
+# Admin App custom widget editor
+
+With this AdminApp extensibility point, you can create your own cusom widget designers and override the defualt auto generated designers.
```
Comment (Novkirishki) [inline src/widget-editor/ReadMe.md:3]: default

File: src/widget-editor/ReadMe.md [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+In this example we have two provided two samples:
```
Comment (Novkirishki) [inline src/widget-editor/ReadMe.md:5]: double two

File: src/widget-editor/ReadMe.md [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+* For overriding a specific field for a widget designer.
+
+* For overriding the entire view for a widget designer.
+
+
+## Overriding a custom field
+
+The process is exactly the same as it is with overriding a field for a content type with the exception that the typeName of the item is equal to widget-{WIDGET_NAME}.
+
+The process invloves registering a custom field provider class by inheriting from the FieldsProvider provider interface and implementing the method "overrideField". This method is called for each and every field when it is beeing rendered in the interface. Thus you are able to replace each field.
```
Comment (Novkirishki) [inline src/widget-editor/ReadMe.md:16]: Lets add link to the API docs as in the other readme files. For example **FieldsProvider** to be a link to: http://admin-app-extensions-docs.sitefinity.site/interfaces/fieldsprovider.html
overrideFields -> http://admin-app-extensions-docs.sitefinity.site/interfaces/fieldsprovider.html#overridefield
etc.

File: src/widget-editor/ReadMe.md [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+With this AdminApp extensibility point, you can create your own cusom widget designers and override the defualt auto generated designers.
+
+In this example we have two provided two samples:
+
+* For overriding a specific field for a widget designer.
+
+* For overriding the entire view for a widget designer.
+
+
+## Overriding a custom field
```
Comment (Novkirishki) [inline src/widget-editor/ReadMe.md:12]: double new line above

File: src/widget-editor/ReadMe.md [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+* fieldType -> The type of the field(shorttext, longtext, for all available field types see the FieldTypes class)
+* fieldName -> Holds the name of the field (Title, Content, Description)
+* typeName -> Holds the type, whose fields are currently beeing rendered.
+
+When a widget designer is opened, the typeName property will have the value of widget-{WIDGET_NAME}, where WIDGET_NAME is the name of the widget and you can choose to override the fields for each widget selectevely.
+
+For reference please refer to the [custom fields sample](../custom-fields/README.md) and the file "widget-editor-custom-fields-provider".
+
+
+## Overriding the entire designer
```
Comment (Novkirishki) [inline src/widget-editor/ReadMe.md:29]: double new line above

File: src/widget-editor/ReadMe.md [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+If you wish to replace the appearance of the entire designer, then you can do so by inheriting from the WidgetEditorViewProvider interface. This interface gets called each time a widget designer is opened. It contains a single method(overrideView) which gets invoked by the AdminApp. This method gets invoked with a single parameter(context) that has the property widgetName. The name of the currently edited widget is passed as a value in this parameter. Thus you can selectevely deside whether to override the designer for a specific widget or not. For reference see the [CustomWidgetEditorViewProvider](custom-editor/custom-widget-editor-view-provider.ts) class.
```
Comment (Novkirishki) [inline src/widget-editor/ReadMe.md:31]: links to the API docs again (I suppose we would need to generate and update them)

File: src/widget-editor/custom-editor/custom-widget-editor-view-provider.ts [hunk 7]
```diff
@@ -0,0 +1,28 @@
+import { Injectable, ClassProvider } from "@angular/core";
+import { EditorComponentContext, EditorViewComponent, WidgetEditorViewProvider, WIDGET_VIEW_TOKEN } from "progress-sitefinity-adminapp-sdk/app/api/v1";
+import { CustomWidgetEditorComponent } from './custom-widget-editor.component';
+
+/**
+ * The ViewRegistrationProvider class works with all of the edior view providers and combines their views into a single
```
Comment (Novkirishki) [inline src/widget-editor/custom-editor/custom-widget-editor-view-provider.ts:6]: I do not understand this comment

File: src/widget-editor/custom-editor/custom-widget-editor.component.ts [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+    get propertiesValue() {
+        return JSON.stringify(this._value, null, 2);
+    }
+
+    set propertiesValue(v) {
+        try {
+            this._value = JSON.parse(v);
+            this.detector.detectChanges();
+        } catch (e) {
+            console.log("error occored while you were typing the JSON");
```
Comment (jmeleshko) [inline src/widget-editor/custom-editor/custom-widget-editor.component.ts:61]: occurred

File: src/widget-editor/ReadMe.md (hunk 1 above)
Comment (jmeleshko) [reply src/widget-editor/ReadMe.md:3]: also cusom -> custom

File: src/widget-editor/ReadMe.md (hunk 3 above)
Comment (jmeleshko) [inline src/widget-editor/ReadMe.md:16]: invloves -> involves

Comment (jmeleshko) [reply src/widget-editor/ReadMe.md:16]: beeing -> being

File: src/widget-editor/ReadMe.md [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+
+The process is exactly the same as it is with overriding a field for a content type with the exception that the typeName of the item is equal to widget-{WIDGET_NAME}.
+
+The process invloves registering a custom field provider class by inheriting from the FieldsProvider provider interface and implementing the method "overrideField". This method is called for each and every field when it is beeing rendered in the interface. Thus you are able to replace each field.
+
+ The parameter that is passed to the method contains three properties:
+
+* fieldType -> The type of the field(shorttext, longtext, for all available field types see the FieldTypes class)
+* fieldName -> Holds the name of the field (Title, Content, Description)
+* typeName -> Holds the type, whose fields are currently beeing rendered.
```
Comment (jmeleshko) [inline src/widget-editor/ReadMe.md:22]: beeing -> being

File: src/widget-editor/ReadMe.md [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+
+When a widget designer is opened, the typeName property will have the value of widget-{WIDGET_NAME}, where WIDGET_NAME is the name of the widget and you can choose to override the fields for each widget selectevely.
```
Comment (jmeleshko) [inline src/widget-editor/ReadMe.md:24]: selectevely -> selectively

File: src/widget-editor/ReadMe.md (hunk 6 above)
Comment (jmeleshko) [reply src/widget-editor/ReadMe.md:31]: selectevely -> selectively

Comment (jmeleshko) [reply src/widget-editor/ReadMe.md:31]: deside -> decide

File: src/widget-editor/ReadMe.md [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+
+
+## Overriding the entire designer
+
+If you wish to replace the appearance of the entire designer, then you can do so by inheriting from the WidgetEditorViewProvider interface. This interface gets called each time a widget designer is opened. It contains a single method(overrideView) which gets invoked by the AdminApp. This method gets invoked with a single parameter(context) that has the property widgetName. The name of the currently edited widget is passed as a value in this parameter. Thus you can selectevely deside whether to override the designer for a specific widget or not. For reference see the [CustomWidgetEditorViewProvider](custom-editor/custom-widget-editor-view-provider.ts) class.
+
+As a return result from this method, it is required to return the reference to the component that you wish to be displayed. This component must implement the interface WidgetEditor. It contains the following methods:
+
+* initialize -> invoked when the custom editor is instantiated for any kind of initialization logic. As a parameter we pass all the widge and property metadata that can be used for generating the UI.
```
Comment (jmeleshko) [inline src/widget-editor/ReadMe.md:35]: widge -> widget

File: src/widget-editor/ReadMe.md [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+
+* setValues -> invoked when the property values for the designer are retrieved from the server and here they are passed as a parameter in order to set them to the fields.
+
+* validate -> invoked when save is performed. Here we require to validate the designer and if there are errors present return them as a return paremeter.
```
Comment (jmeleshko) [inline src/widget-editor/ReadMe.md:39]: paremeter -> parameter

File: src/widget-editor/ReadMe.md [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+
+* actionExecuting -> invoked when an action is executed (save) so you can plug-in your own custom logic.
+
+* getModifiedProperties -> invoked to return the modified properties from the desigener.
```
Comment (jmeleshko) [inline src/widget-editor/ReadMe.md:43]: desigener -> designer

**Review by Novkirishki — APPROVED** [review]

**Review by jmeleshko — APPROVED** [review]

**Pull request merged**